flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/DenseVector.scala

```diff
@@ -102,6 +102,38 @@ case class DenseVector(
     }
   }
 
+  /** Returns the outer product (a.k.a. Kronecker product) of `this`
+    * with `other`. The result will given in [[org.apache.flink.ml.math.SparseMatrix]]
+    * representation if `other` is sparse and as [[org.apache.flink.ml.math.DenseMatrix]] otherwise.
+    *
+    * @param other a Vector
+    * @return the [[org.apache.flink.ml.math.Matrix]] which equals the outer product of `this`
+    *         with `other.`
+    */
+  override def outer(other: Vector): Matrix = {
+    val numRows = size
+    val numCols = other.size
+
+    other match {
+      case SparseVector(size, indices, data_) =>
+        val entries: Array[(Int, Int, Double)] = for {
+          i <- (0 until numRows).toArray
+          j <- indices
+          value = this(i) * other(j)
```

Comment:
It might make sense to work directly on the `data` array here, because every `other(j)` call entails a binary search operation. If you zip `data` with `indices`, then you should have all information necessary to access `this(i)` and to have the value for `other(j)`.